unix: Add a modbluetooth implementation supporting USB adaptors.

jimmo opened the pull request:
This PR adds a "btstack" variant that links in btstack and provides an implementation of the port-specific functionality using the "libusb" btstack port.

Some improvements to the way stack initialisation and deinitialisation were added to modbluetooth_btstack, which should be beneficial for STM32 also.

I've been testing this with a BCM20702A-based USB adaptor (specifically https://au.targus.com/products/bluetooth-4-0-dual-mode-micro-usb-adapter-acb75au). This PR passes the multitests with the six combinations of PYBD-NimBLE (a), PYBD-BTStack (b), Unix-BTStack (c) (ab, ba, bc, cb, ac, ca). (BTW, the multitests were extremely useful in getting this working!).

In theory this can be easily adapted to other btstack ports, of which the libusb-intel, raspi and various posix-h4/h5 are probably the most interesting. It would also be good to make the USB device path configurable via `hci_transport_usb_set_path`, which would also allow the multi-tests to run with two adaptors on the same machine.

The filtering for run-multitests.py isn't ideal, open to suggestions there (maybe the test .py or .exp could specify these). I'll try and figure out a way to suppress these messages from btstack/libusb.

Note this PR requires #5859 (as it uses the scheduler from multiple threads), and includes the commits from that PR. It also needs bluekitchen/btstack#278.

@mringwal FYI

jimmo committed:
unix/modmachine.c: Add machine.idle() (using sched_yield).

jimmo committed:
stm32/Makefile: Rename SRC_LIB to LIB_SRC_C to match other ports.

jimmo committed:
extmod/modbluetooth: Don't hold atomic section for call to mp_sched_schedule.

jimmo committed:
stm32/main.c: Clean shutdown of btstack.

jimmo committed:
extmod/btstack: More robust init/deinit sequencing.

jimmo committed:
unix: Add support for modbluetooth using btstack.

Uses the libusb HCI transport (supports many common USB BT adaptors).

jimmo committed:
tests/run-multitests.py: Allow filtering out lines from stdout.

jimmo committed:
tests/multi_bluetooth/ble_gap_advertise: Fix bytes/str comparison warning.

Also ignore scan response events for now (need to make this behavior consistent).

jimmo committed:
tests/run-multitests.py: Add shortcuts for local Python instances.

`-i micropython` and `-i cpython` to add instances using the `MICROPYTHON` and
`CPYTHON3` variables (which can be overridden by env vars).

jimmo commented:
@stinos This breaks the Windows build because I used `sched_yield` to implement `machine.idle()`.

I'm not 100% sure that sched_yield is even the right thing to use here, but mp_hal_delay_us (or ms) feels wrong too. Should I make this a macro and allow Windows to override it (do you have a suggested equivalent, maybe `SwitchToThread()` ?).

Maybe adding machine.idle() to the unix port doesn't make sense anyway. The multitests use it while they wait for events to occur, which might actually be better to use time.sleep_ms. But idle() is what I'd suggest someone use if they were writing this code on a microcontroller.

Open to suggestions!

jimmo commented:
I spoke to @stinos and we decided to disable machine.idle for Windows, which I have now done by making mpconfigport.h able to define a yield implementation.

Longer term obviously the ideal would be to make this all select() driven, but btstack currently doesn't have a way to select() on libusb, and we'd have to integrate this somehow with stdio etc and all other ways that would make this equivalent to WFI on the baremetal ports.

dpgeorge commented on tests/run-tests:
```diff
@@ -354,6 +354,7 @@ def run_tests(pyb, tests, args, base_path="."):
 
     if not has_coverage:
         skip_tests.add('cmdline/cmd_parsetree.py')
+        skip_tests.add('unix/schedule_thread.py') # Needs assert()
```
This test can still run on any unix variant, right?  Even without assert() enabled in the executable it will still stress test schedule+thread.

And looking at the test itself, seems it just needs threading and scheduler, so could probably be named `tests/thread/stress_schedule.py` to run on any target with threading (eg esp32).

jimmo replied:
Yes, it will run on any build that has _thread and schedule. However, the issue is that the test only fails by triggering the assert. Without the assert the most likely failure mode is an infinite loop as the scheduler stops scheduling. I guess the test could add its own timeout to the main loop?

dpgeorge replied:
> However, the issue is that the test only fails by triggering the assert. Without the assert the most likely failure mode is an infinite loop as the scheduler stops scheduling.

But at least that's showing a bug.  I think it's important to be able to run this stress test on any port that it can run on.

> I guess the test could add its own timeout to the main loop?

If that's possible, yes.

dpgeorge commented on tests/run-multitests.py:
```diff
@@ -394,6 +403,8 @@ def main():
         if i.startswith("exec:"):
             instances_test.append(PyInstanceSubProcess([i[len("exec:") :]]))
+        elif i == "unix":
+            instances_test.append(PyInstanceSubProcess([MICROPYTHON]))
```
This will work also on windows (and other systems), so maybe it should be called something else... but something like "local_micropython" is long.  Could also add "cpython" because it can be useful to run against CPython sometimes.

jimmo replied:
how about "micropython" and "cpython" ? (which matches the names of the constants anyway)

dpgeorge replied:
Yeah, that makes the most sense.  Could also have "micropython-coverage" then.

dpgeorge commented on tests/multi_bluetooth/ble_gap_advertise.py:
```diff
@@ -37,18 +37,21 @@ def irq(ev, data):
-                else:
-                    if adv_data != data[4]:
-                        adv_data = "MISMATCH"
+                if data[2] in (0, 2):
```
This (getting scan responses) was exactly the issue I originally had with BTstack on PYBD, which is now fixed by the latest BTstack update.  Did you use the latest master?  Or was there some other reason scan responses are passed through to here?

jimmo replied:
Yes, latest master, but I may have forgotten to update the submodule. I'll check tomorrow and if that's not it I'll find out for sure.

dpgeorge commented:
> This PR adds a "btstack" variant

I'm not sure if a new variant is needed: I would think that the coverage variant would have btstack enabled to be able to see coverage of the bluetooth module by the multi-tests.  And the dev variant would also have it enabled for development with bluetooth.  But maybe there's a good reason not to have bluetooth by default?

> This PR passes the multitests with the six combinations of ...

Good to hear!  Actually I wanted to add a `--symmetry` option to the multi-test runner that would try all permutations of instances.  For only 2 instances it would test AB and BA if A!=B.  But that's something to add in the future.

jimmo commented:
> I'm not sure if a new variant is needed: I would think that the coverage variant would have btstack enabled to be able to see coverage of the bluetooth module by the multi-tests. And the dev variant would also have it enabled for development with bluetooth. But maybe there's a good reason not to have bluetooth by default?

Yeah, no strong feelings on this. And it would be good to do coverage.

OK, I'll match STM32 then -- a single makefile flag `MICROPY_PY_BLUETOOTH` that defaults to off that can be set by the variant or make command line (and set this flag in dev & coverage).

MICROPY_PY_BLUETOOTH would imply MICROPY_BLUETOOTH_BTSTACK and MICROPY_BLUETOOTH_BTSTACK_USB.

> Actually I wanted to add a `--symmetry` option to the multi-test runner that would try all permutations of instances.

Yes! I nearly did this myself. Definitely useful in the future.

dpgeorge commented:
> OK, I'll match STM32 then -- a single makefile flag `MICROPY_PY_BLUETOOTH` that defaults to off that can be set by the variant or make command line (and set this flag in dev & coverage).

Sounds good.

jimmo committed:
unix: Enable modbluetooth on the "dev" and "coverage" variants.

jimmo commented:
Have addressed the comments above -- no longer a variant, enabled in dev and coverage, confirmed that updating the btstack submodule solved the scan_resp issue, renamed the shortcut for run-multitest.py.

I'll address the scheduler stress test in #5859

mringwal commented:
> Longer term obviously the ideal would be to make this all select() driven, but btstack currently doesn't have a way to select() on libusb, 

Would using select in BTstack's libusb driver help much? The polling code is quite simple, too.

The code to use select() is there, it's just commented out as it didn't work on mac and linux according to the comment (and I don't remember the details). From the libusb specs, it should work on both (darwin and newer linux than 2.6.27 ) - see http://libusb.sourceforge.net/api-1.0/group__libusb__poll.html#libusb_pollmain

dpgeorge commented:
Can you please rebase now that #5859 is in?

jimmo commented:
Done.

mringwal commented:
When compiling on macOS with Xcode's clang, I get this errors:

```
CC ../../extmod/btstack/modbluetooth_btstack.c
../../extmod/btstack/modbluetooth_btstack.c:46:82: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
STATIC btstack_packet_callback_registration_t hci_event_callback_registration = {0};
```

```
CC ../../extmod/btstack/modbluetooth_btstack.c
../../extmod/btstack/modbluetooth_btstack.c:223:62: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
STATIC btstack_timer_source_t btstack_init_deinit_timeout = {0};
```

mringwal commented:
How can I try this? I've checked out this branch and used
`make VARIANT=dev` to compile in ports/unix. When starting mp, I get:
```
 ./micropython
MicroPython v1.12-204-g91ddc6bbe-dirty on 2020-02-23; darwin version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import bluetooth
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'bluetooth'
>>> 
```

jimmo commented:
> When compiling on macOS with Xcode's clang, I get this errors:

I pushed a commit to remove these initialisers.

(And you need to use ./micropython-dev for the dev variant, looks like you figured that out though)

mringwal commented:
> (And you need to use ./micropython-dev for the dev variant, looks like you figured that out though)
Thanks, I eventually had a look at the version/date reported by MP on start and after some wondering (make clean ; make ) realized that there's only a -dev variant, which supports `import bluetooth`.

jimmo committed:
lib/btstack: Update to c8b9823for USB HCI reset timeout fix.

mringwal commented:
Hi. Please update to btstack develop again, there was a regression that prevented outgoing classic connections (oops..)

jimmo committed:
extmod/modbluetooth: Fix warnings (sign compare, unused).

jimmo committed:
unix: Add btstack to the unix submodules.

Only when bluetooth is enabled, i.e. if building the dev or coverage variants,
and we have libusb available.

Update travis to match (i.e. specify the variant when doing `make submodules`).

jimmo commented:
Updated the Makefiles and Travis config to get the tests passing by making the modbluetooth support additionally conditional on having libusb available.

dpgeorge commented on tests/multi_bluetooth/ble_gap_advertise.py:
```diff
@@ -42,13 +42,13 @@ def irq(ev, data):
     ble.config(rxbuf=2000)
     ble.irq(irq)
-    ble.gap_scan(2 * ADV_TIME_S * 1000, 10000, 10000)
+    ble.gap_scan(2 * ADV_TIME_S * 1000, 30000, 30000)
```
Is this change still needed now that the scan-response data issue is fixed?

jimmo replied:
Oh nope, I had it in my head that 30000 is the minimum (it isn't, rather it's NimBLE's default for "fast" scanning). Removed this change.

dpgeorge commented on extmod/btstack/btstack.mk:
```diff
@@ -24,25 +28,30 @@ INC += -I$(BTSTACK_DIR)/3rd-party/yxml
+# Suppress some warnings (e.g. %lu vs %u, unused variables, etc).
+BTSTACK_WARNING_CFLAGS = -Wno-error=old-style-definition -Wno-error=unused-variable -Wno-error=unused-parameter
+ifneq ($(CC),clang)
+BTSTACK_WARNING_CFLAGS += -Wno-error=format=
```
I didn't see a need for this, not in unix coverage, dev or 32-bit coverage.  Do you know where it's needed?

jimmo replied:
It's needed for a few things in the BT Classic code, and also for le_device_db_tlv.c

Neither are currently included in the build, but I'll likely have to add this back later.

dpgeorge replied:
Ok!

dpgeorge commented on extmod/btstack/btstack.mk:
```diff
@@ -24,25 +28,30 @@ INC += -I$(BTSTACK_DIR)/3rd-party/yxml
+LIB_SRC_C += $(SRC_BTSTACK)
 
+# Suppress some warnings (e.g. %lu vs %u, unused variables, etc).
+BTSTACK_WARNING_CFLAGS = -Wno-error=old-style-definition -Wno-error=unused-variable -Wno-error=unused-parameter
```
I'd prefer to have these as fully disabling the warning, rather than the error due to the warning, so that the output of the build process is clean (ie no warnings).  So just `-Wno-old-style-definition -Wno-unused-variable -Wno-unused-parameter`

jimmo replied:
Fixed.

dpgeorge commented:
Re `machine.idle()` on unix: yes the multi-bluetooth tests could easily be changed to use a delay instead of this idle call, but I think it's nice to add `machine.idle()` to the unix port, so this is a good change.

For the commit "extmod/modbluetooth: Don't hold atomic section for call to mp_sched_schedule.", this looks good but can you explain the reason for it?

The setting `MICROPY_PY_URANDOM_EXTRA_FUNCS` is enabled in the unix dev variant, was this intentional?  Of course it'd be good to enable this but not in this set of commits if it's not needed.

I think "extmod/modbluetooth_btstack: Remove unnecessary init of btstack registrations." should be squashed into "extmod/btstack: More robust init/deinit sequencing." where those `={0}`'s were added.

Apart from that, this looks good to go in.

jimmo commented:
> For the commit "extmod/modbluetooth: Don't hold atomic section for call to mp_sched_schedule.", this looks good but can you explain the reason for it?

On bare-metal this is fine, but on unix the atomic section isn't re-entrant. And sched_schedule will re-acquire the atomic section.

> 
> The setting `MICROPY_PY_URANDOM_EXTRA_FUNCS` is enabled in the unix dev variant, was this intentional? Of course it'd be good to enable this but not in this set of commits if it's not needed.

The multitests use it.

> 
> I think "extmod/modbluetooth_btstack: Remove unnecessary init of btstack registrations." should be squashed into "extmod/btstack: More robust init/deinit sequencing." where those `={0}`'s were added.

Agreed will do that now.

dpgeorge commented:
> The multitests use it.

Which one exactly?

jimmo commented:
> > The multitests use it.
> 
> Which one exactly?

Oh sorry. The BLE examples use it. ble_temperature.py uses random.uniform.

dpgeorge commented on extmod/modbluetooth.c:
```diff
@@ -918,11 +935,15 @@ STATIC bool enqueue_irq(mp_obj_bluetooth_ble_t *o, size_t len, uint16_t event) {
 }
 
-STATIC void schedule_ringbuf(void) {
+// Must not hold the atomic section.
```
shouldn't this be the opposite, "Must hold the Bluetooth atomic lock upon entry to this function."?

jimmo replied:
Done.

dpgeorge commented:
Not sure why PYBD-SF2 build is failing now with a NimBLE link error...

dpgeorge commented:
> Not sure why PYBD-SF2 build is failing now with a NimBLE link error...

Looks like a change to the nimble submodule was inadvertently included in a commit here.  I'll undo that during merge.

dpgeorge commented:
This was merged in cb5994d96e45f26fab4abf5b52ea0f419a060641 through 2e3c42775a402749e790474643cb9f776561264d

Thanks a lot, this is a great addition!
